pkg/controller/mizar/mizar-node-controller.go

```diff
@@ -0,0 +1,293 @@
+}
+
+// Enqueue puts key of the node object in the work queue
+// EventType: Create=0, Update=1, Delete=2, Resume=3
+func (c *MizarNodeController) Enqueue(key string, eventType EventType) {
+	c.queue.Add(KeyWithEventType{Key: key, EventType: eventType})
+}
+
+// Dequeue an item and process it
+func (c *MizarNodeController) runWorker() {
+	for {
+		item, queueIsEmpty := c.queue.Get()
+		if queueIsEmpty {
+			break
+		}
+		c.process(item)
+	}
+}
+
+// Parsing a item key and call gRPC request
+func (c *MizarNodeController) process(item interface{}) {
+	defer c.queue.Done(item)
+	keyWithEventType, ok := item.(KeyWithEventType)
+	if !ok {
+		klog.Errorf(c.logMessage("Unexpected item in queue: ", keyWithEventType))
+		c.queue.Forget(item)
+		return
+	}
+
+	key := keyWithEventType.Key
+	eventType := keyWithEventType.EventType
+	nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress, err := c.parseKey(key)
+	if err != nil || nodeKind != NodeKind {
+		klog.Errorf(c.logMessage("Unexpected string in queue; discarding: ", key))
+		c.queue.Forget(item)
+		return
+	}
+	klog.Infof(c.logInfoMessage("Processing a node: ")+"%s/%s/%s/%s/%s/%s/%s", nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress)
+	result, err := c.gRPCRequest(eventType, nodeName, nodeAddress)
+	if !result {
+		klog.Errorf(c.logMessage("Failed a node processing: ", key))
+		c.queue.Add(item)
+	} else {
+		klog.Infof(c.logMessage(" Processed a node: ", key))
+		c.queue.Forget(item)
+	}
+}
+
+// Generate a node key
+func (c *MizarNodeController) genKey(resource *v1.Node) string {
+	nodeKind := resource.GetObjectKind().GroupVersionKind().Kind
+	if nodeKind == "" {
+		nodeKind = NodeKind
+	}
+	nodeTenant := resource.GetTenant()
+	nodeName := resource.GetName()
+	if nodeName == "" {
+		klog.Fatalln(c.logMessage("Node name should not be null", nodeName))
+	}
+	var nodeStatusType, nodeStatus string
+	conditions := resource.Status.Conditions
+	if len(conditions) < 6 {
+		klog.Fatalln(c.logInfoMessage("Node status information is needed"))
+	} else {
+		nodeStatusType = fmt.Sprintf("%s", resource.Status.Conditions[5].Type)
+		nodeStatus = fmt.Sprintf("%s", resource.Status.Conditions[5].Status)
+	}
+	var nodeAddressType, nodeAddress string
+	addresses := resource.Status.Addresses
+	if len(addresses) < 1 {
+		klog.Fatalln(c.logInfoMessage("Node ip address is needed"))
+	} else {
+		nodeAddressType = fmt.Sprintf("%s", resource.Status.Addresses[0].Type)
+		nodeAddress = fmt.Sprintf("%s", resource.Status.Addresses[0].Address)
+	}
+	key := fmt.Sprintf("%s/%s/%s/%s/%s/%s/%s", nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress)
+	return key
+}
+
+// Parse a key and get node information
+func (c *MizarNodeController) parseKey(key string) (nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress string, err error) {
```

Comment:
Why we need so many sections in the key string? Just node name is enough, as each node is uniquely named within a cluster.

Reply:
This was because it can avoid object processing twice to convert Update event to Update or Resume events. Anyway I changed this part because it looks like your suggestion is more common.